### PR TITLE
Improve the VoiceNext queue

### LINKED ISSUE #537: Allow canceling voice transmissions
# Summary
Outgoing VoiceNext transmissions should be cancellable.

# Details
I tried killing the ffmpeg process, flushing and disposing the transmission stream but the only half-working solution I could find was disconnecting and reconnecting the bot from the voice channel; but we need a better way to interrupt the playback and clear the stream.

# Notes
Not sure if these are related but I've noticed, as the audio files get bigger, more and more memory is used and not freed by the GC even after playback is complete and the bot is disconnected from the voice channel. I played around with dotMemory a bit and noticed after disconnecting (disposing the connection) and reconnecting a few times, multiple instances of VoiceNextConnection show up in the memory snapshot, not even sure if this is unintended behaviour..

### PR DESCRIPTION
**Pull request by WamWooWam**
# Summary
Fixes #537 by limiting the size of the VoiceNext packet queue, and modifying the queue to store PCM instead of encoded Opus.

# Details
I've switched out the `ConcurrentQueue` for a `System.Threading.Channels.Channel` to limit number of packets that can be in the queue at any one time, defaulting to 25 packets or about 500ms of audio. This means that you have control over the buffer within VoiceNext, and also means that when you stop writing data, the audio will actually stop on Discord's end too. You can now pause audio! 🎉

This means that the entire VoiceNext pipeline is now asynchronous, which unfortunately doesn't play nicely with `System.IO.Stream`, and as such [we decided it would be best to refactor this](https://discord.com/channels/379378609942560770/379386901725052928/777152309901262859). As such, `VoiceTransmitStream` is no longer a subclass of `System.IO.Stream`, and has been renamed to `VoiceTransmitSink` to avoid confusion, but the functionality is much the same. That said, **this is a breaking change**, as you can no longer synchronously send data to VoiceNext.

I also made the queue store PCM audio, instead of compressed Opus, which potentially fixes some threading issues (OpusEncoders aren't thread safe) and also spreads out the CPU load of encoding packets to the time they're actually sent. And let's be real, if you can't encode 20ms of Opus audio in under 20ms, something's gone wrong.

Oh I also fixed UdpPing, because it would always return zero.

This also reduces the memory usage of VoiceNext by quite a fair bit, even if we are storing PCM data.

# Changes proposed
* Rename `VoiceTransmitStream` to `VoiceTransmitSink`
* Limit the size of the VoiceNext queue
* Make the VoiceNext queue store PCM data.
* Add an extension method to replicate `CopyToAsync` functionality on `VoiceTransmitSink`
* Fix the test bot, because capturing stderr then not reading it causes ffmpeg to get quite sad
* Fix the `UdpPing` property on `VoiceNextConnection` always returning 0.
* Repeat myself adnauseam.

### CODE REVIEW

**Comment by Giggitybyte** [comment]
Based Wam

**Comment by Emzi0767** [comment]
If you're limiting queue size, why not use a bounded channel instead? It's async, so won't be stalling tasks.

Sent from Outlook Mobile<https://aka.ms/blhgte>

________________________________

**Comment by WamWooWam** [comment]
would System.Threading.Channels not be another dependency and that's bad?

**Comment by Emzi0767** [comment]
Meh. We can afford it.

**Comment by WamWooWam** [comment]
The problem with this is that you could then only write to the stream asynchronously, while as it stands all the code we have depends on synchronous writing, on top of that we're already blocking on tasks no matter what we really do because of the way the send buffer works

**Comment by Emzi0767** [comment]
Are you sure about that?

NuGet says otherwise

**Comment by WamWooWam** [comment]
More, since when have we been able to override `Stream#WriteAsync`

File: DSharpPlus.VoiceNext/VoiceNextConnection.cs [hunk 1]
```diff
@@ -375,7 +377,16 @@ private async Task VoiceSenderTask()
                 // time.time()
                 //   DateTime.Now
 
-                var durationModifier = hasPacket ? packet.MillisecondDuration / 5 : 4;
+                if (hasPacket) {
```
Comment (Emzi0767) [inline DSharpPlus.VoiceNext/VoiceNextConnection.cs:380]: Brace style.

**Review by Emzi0767 — CHANGES_REQUESTED** [review]

**Comment by Emzi0767** [comment]
So thing with BlockingCollection is... the fact that it blocks. And that stalls tasks. And that starves the scheduler's pool. This is why I'd prefer an async-based (e.g. Channels) solution.

**Comment by WamWooWam** [comment]
As per discussion [in the Discord](https://discord.com/channels/379378609942560770/379386901725052928/777152309901262859), I've refactored `VoiceTransmitStream` to no longer be a subclass of `System.IO.Stream` because it doesn't provide a good interface for async-only IO.

`VoiceTransmitStream` has been renamed to `VoiceTransmitSink` to avoid confusion, all Read/Seek/Length methods have been removed and an extension method has been provided to make copying from `IO.Stream`s as simple as possible.

**Comment by Emzi0767** [comment]
It seems the natives were committed?

**Review by Emzi0767 — APPROVED** [review]

**Comment by WamWooWam** [comment]
This also should close #481, because any functionality that PR could provide this one provides implicitly

**Pull request merged**